### PR TITLE
[ci] fixed newline issue and changed cpplint code to check number of issues

### PR DESCRIPTION
**Pull request by jameslamb**
@StrikerRUS when I went to update #1990 after #2563 was merged, I found that we still had 87 `cpplint` issues (I had expected it to drop to 86). I think it's possible that a new one was introduced in a recent PR.

In this PR, I fix that issue but also propose that we change our approach to `cpplint`. Instead of saying "we run it to get the logs but it doesn't fail builds", I propose that we do "we track how many `cpplint` issues remain and fail builds that introduce _new_ ones"

### CODE REVIEW

**Comment by StrikerRUS** [comment]
Cool idea! Please exclude `lint` job from macOS by updating this line:
https://github.com/microsoft/LightGBM/blob/1f7b06b98391395b6fd2f0bd7937353254718956/.travis.yml#L34-L35

**Comment by jameslamb** [comment]
> Cool idea! Please exclude `lint` job from macOS by updating this line:
> https://github.com/microsoft/LightGBM/blob/1f7b06b98391395b6fd2f0bd7937353254718956/.travis.yml#L34-L35

ah, got it! I forgot about that, thank you

**Comment by jameslamb** [comment]
It seems unlikely that this PR's changes are causing [this Appveyor failure](https://ci.appveyor.com/project/guolinke/lightgbm/builds/28913296/job/nm5kl94cwett174c), but I don't think I have sufficient permissions to trigger a rebuild myself.

**Comment by StrikerRUS** [comment]
> It seems unlikely that this PR's changes are causing this Appveyor failure, but I don't think I have sufficient permissions to trigger a rebuild myself.

You can close and re-open again to re-trigger all CI builds for PRs (unfortunately, not for branches). Exactly that I'm going to do right now :-).

**Comment by jameslamb** [comment]
> > It seems unlikely that this PR's changes are causing this Appveyor failure, but I don't think I have sufficient permissions to trigger a rebuild myself.
> 
> You can close and re-open again to re-trigger all CI builds for PRs (unfortunately, not for branches). Exactly that I'm going to do right now :-).

ah right, I always forget about that. Thanks! Seems the checks are passing now.

File: .ci/test.sh [hunk 1]
```diff
@@ -65,7 +65,21 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "lint" ]]; then
     echo "Linting R code"
     Rscript ${BUILD_DIRECTORY}/.ci/lint_r_code.R ${BUILD_DIRECTORY} || exit -1
     echo "Linting C++ code"
-    cpplint --filter=-build/c++11,-build/include_subdir,-build/header_guard,-whitespace/line_length --recursive ./src ./include || exit 0
+    cpplint \
+        --filter=-build/c++11,-build/include_subdir,-build/header_guard,-whitespace/line_length \
+        --recursive \
+        ./src \
+        ./include 2>&1 | tee cpplint.txt | cat \
```
Comment (StrikerRUS) [inline .ci/test.sh]: Please include this file into `.gitignore`.

**Comment by jameslamb** [comment]
> I'm a little worried about this HARD restrictions will slow down the PR. @guolinke, what're your comments?

in my opinion, the slow down is a good thing. It means we take this linting stuff and spread out the effort in tiny things over many PRs, instead of needing separate linting-fix-only PRs after the fact.

I also think having this set up means we could make a decision like "I am getting a new linting error and don't know how to fix it right now so I'm increasing `MAX_CPPLINT_ERRORS` by 1" and it would be clear in the diff that that had happened

**Comment by StrikerRUS** [comment]
@jameslamb Please update the branch - we have `Total errors found: 97` in `master` 😢 .

**Comment by jameslamb** [comment]
I rebased to `master` and found 91 errors

<details><summary>logs (click me)</summary>

```
src/io/json11.cpp:53:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:57:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:67:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:73:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:77:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:114:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:126:  Is this a non-const reference? If so, make const or use a pointer: string &out  [runtime/references] [2]
src/io/json11.cpp:339:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
src/io/json11.cpp:351:  const string& members are dangerous. It is much better to use alternatives, such as pointers or simple constants.  [runtime/member_string_references] [2]
src/io/json11.cpp:452:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
src/io/json11.cpp:452:  Is this a non-const reference? If so, make const or use a pointer: string & out  [runtime/references] [2]
src/io/json11.cpp:479:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
src/io/json11.cpp:523:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
src/network/socket_wrapper.hpp:221:  Are you taking an address of a cast?  This is dangerous: could be a temp var.  Take the address before doing the cast, rather than after  [runtime/casting] [4]
src/io/dataset.cpp:30:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/dataset.cpp:341:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/main.cpp:8:  Found C++ system header after other header. Should be: main.h, c system, c++ system, other.  [build/include_order] [4]
src/main.cpp:41:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
src/metric/dcg_calculator.cpp:27:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/treelearner/feature_histogram.hpp:676:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
include/LightGBM/json11.hpp:80:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:81:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:82:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:83:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:84:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:85:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:86:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:87:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:88:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:89:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:90:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:94:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:101:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:107:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:111:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
include/LightGBM/json11.hpp:144:  Is this a non-const reference? If so, make const or use a pointer: std::string &out  [runtime/references] [2]
include/LightGBM/json11.hpp:153:  Is this a non-const reference? If so, make const or use a pointer: std::string & err  [runtime/references] [2]
include/LightGBM/json11.hpp:156:  Is this a non-const reference? If so, make const or use a pointer: std::string & err  [runtime/references] [2]
include/LightGBM/json11.hpp:168:  Is this a non-const reference? If so, make const or use a pointer: std::string::size_type & parser_stop_pos  [runtime/references] [2]
include/LightGBM/json11.hpp:169:  Is this a non-const reference? If so, make const or use a pointer: std::string & err  [runtime/references] [2]
include/LightGBM/json11.hpp:174:  Is this a non-const reference? If so, make const or use a pointer: std::string & err  [runtime/references] [2]
include/LightGBM/json11.hpp:193:  Is this a non-const reference? If so, make const or use a pointer: std::string & err  [runtime/references] [2]
include/LightGBM/json11.hpp:208:  Is this a non-const reference? If so, make const or use a pointer: std::string &out  [runtime/references] [2]
src/c_api.cpp:421:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
src/c_api.cpp:920:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/bin.cpp:80:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/bin.cpp:511:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/lightgbm_R.cpp:36:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
src/io/config_auto.cpp:298:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/io/config_auto.cpp:304:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:305:  Consider using CHECK_LE instead of CHECK(a <= b)  [readability/check] [2]
src/io/config_auto.cpp:312:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/io/config_auto.cpp:383:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:386:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:395:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:398:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:433:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:440:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:443:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:516:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:544:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:553:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:562:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:565:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:568:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/config_auto.cpp:571:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/boosting/gbdt.cpp:299:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/treelearner/gpu_tree_learner.h:37:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
include/LightGBM/c_api.h:1041:  Almost always, snprintf is better than strcpy  [runtime/printf] [4]
include/LightGBM/utils/text_reader.h:127:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
include/LightGBM/utils/text_reader.h:288:  { should almost always be at the end of the previous line  [whitespace/braces] [4]
src/boosting/gbdt.h:26:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
src/treelearner/serial_tree_learner.cpp:290:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/treelearner/serial_tree_learner.cpp:302:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/treelearner/serial_tree_learner.cpp:314:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/treelearner/serial_tree_learner.cpp:876:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/metric/regression_metric.hpp:274:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/metric/regression_metric.hpp:296:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
include/LightGBM/utils/common.h:303:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
include/LightGBM/utils/common.h:374:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
include/LightGBM/utils/common.h:419:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
src/treelearner/serial_tree_learner.h:32:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
src/application/predictor.hpp:45:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/application/predictor.hpp:46:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
include/LightGBM/tree.h:487:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
src/network/linkers.h:159:  "private:" should be preceded by a blank line  [whitespace/blank_line] [3]
src/io/dataset_loader.cpp:16:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
src/io/dataset_loader.cpp:440:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/dataset_loader.cpp:592:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
src/io/dataset_loader.cpp:903:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
include/LightGBM/tree_learner.h:16:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
Done processing include/LightGBM/export.h
Done processing src/metric/map_metric.hpp
Done processing include/LightGBM/utils/threading.h
Done processing include/LightGBM/utils/random.h
Done processing src/io/json11.cpp
Done processing src/treelearner/cost_effective_gradient_boosting.hpp
Done processing src/io/dense_nbits_bin.hpp
Done processing src/metric/binary_metric.hpp
Done processing src/application/application.cpp
Done processing src/network/socket_wrapper.hpp
Done processing src/io/dataset.cpp
Done processing src/objective/regression_objective.hpp
Done processing src/network/linkers_socket.cpp
Done processing src/boosting/gbdt_model_text.cpp
Done processing include/LightGBM/application.h
Done processing src/objective/rank_objective.hpp
Done processing src/boosting/gbdt_prediction.cpp
Done processing src/main.cpp
Done processing src/io/parser.hpp
Done processing src/metric/dcg_calculator.cpp
Done processing src/treelearner/data_parallel_tree_learner.cpp
Done processing include/LightGBM/bin.h
Done processing src/boosting/prediction_early_stop.cpp
Done processing src/io/dense_bin.hpp
Done processing src/treelearner/feature_parallel_tree_learner.cpp
Done processing include/LightGBM/dataset_loader.h
Done processing src/boosting/rf.hpp
Done processing include/LightGBM/utils/openmp_wrapper.h
Done processing src/metric/xentropy_metric.hpp
Done processing src/treelearner/gpu_tree_learner.cpp
Done processing include/LightGBM/utils/file_io.h
Done processing src/treelearner/leaf_splits.hpp
Done processing src/treelearner/feature_histogram.hpp
Done processing include/LightGBM/metric.h
Done processing include/LightGBM/json11.hpp
Done processing src/io/metadata.cpp
Done processing src/boosting/boosting.cpp
Done processing src/c_api.cpp
Done processing src/io/bin.cpp
Done processing src/lightgbm_R.cpp
Done processing src/metric/rank_metric.hpp
Done processing src/network/linker_topo.cpp
Done processing include/LightGBM/utils/log.h
Done processing src/network/linkers_mpi.cpp
Done processing src/io/tree.cpp
Done processing src/io/config_auto.cpp
Done processing src/boosting/gbdt.cpp
Done processing src/treelearner/gpu_tree_learner.h
Done processing include/LightGBM/c_api.h
Done processing include/LightGBM/utils/array_args.h
Done processing src/io/sparse_bin.hpp
Done processing include/LightGBM/utils/text_reader.h
Done processing src/boosting/gbdt.h
Done processing src/boosting/score_updater.hpp
Done processing src/treelearner/data_partition.hpp
Done processing src/io/config.cpp
Done processing src/treelearner/serial_tree_learner.cpp
Done processing src/network/network.cpp
Done processing src/treelearner/parallel_tree_learner.h
Done processing src/metric/regression_metric.hpp
Done processing include/LightGBM/dataset.h
Done processing include/LightGBM/lightgbm_R.h
Done processing include/LightGBM/meta.h
Done processing include/LightGBM/utils/pipeline_reader.h
Done processing src/boosting/dart.hpp
Done processing include/LightGBM/config.h
Done processing src/objective/objective_function.cpp
Done processing src/treelearner/voting_parallel_tree_learner.cpp
Done processing include/LightGBM/utils/common.h
Done processing src/io/ordered_sparse_bin.hpp
Done processing src/boosting/goss.hpp
Done processing include/LightGBM/objective_function.h
Done processing src/treelearner/serial_tree_learner.h
Done processing src/metric/multiclass_metric.hpp
Done processing src/objective/binary_objective.hpp
Done processing src/application/predictor.hpp
Done processing include/LightGBM/R_object_helper.h
Done processing include/LightGBM/feature_group.h
Done processing include/LightGBM/prediction_early_stop.h
Done processing include/LightGBM/tree.h
Done processing include/LightGBM/boosting.h
Done processing src/network/linkers.h
Done processing src/objective/multiclass_objective.hpp
Done processing src/io/dataset_loader.cpp
Done processing include/LightGBM/network.h
Done processing src/objective/xentropy_objective.hpp
Done processing src/treelearner/tree_learner.cpp
Done processing src/io/file_io.cpp
Done processing src/treelearner/split_info.hpp
Done processing src/metric/metric.cpp
Done processing include/LightGBM/tree_learner.h
Done processing src/io/parser.cpp
Total errors found: 91
```

</details>

**Comment by guolinke** [comment]
maybe it is better to ignore some types of lint errors, then constraint the error number ?

**Comment by jameslamb** [comment]
> maybe it is better to ignore some types of lint errors, then constraint the error number ?

That is what we are already doing with the `--filter` flag, isn't it?

```
--filter=-build/c++11,-build/include_subdir,-build/header_guard,-whitespace/line_length
```

From `cpplint --help`

```
filter=-x,+y,...
      Specify a comma-separated list of category-filters to apply: only
      error messages whose category names pass the filters will be printed.
```

**Comment by jameslamb** [comment]
> > maybe it is better to ignore some types of lint errors, then constraint the error number ?
> 
> That is what we are already doing with the `--filter` flag, isn't it?
> 
> ```
> --filter=-build/c++11,-build/include_subdir,-build/header_guard,-whitespace/line_length
> ```
> 
> From `cpplint --help`
> 
> ```
> filter=-x,+y,...
>       Specify a comma-separated list of category-filters to apply: only
>       error messages whose category names pass the filters will be printed.
> ```

@guolinke what do you think?

**Pull request closed without merging**